File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s.length() == 0){
            return 0;
        }
        int maxi = INT_MIN;
        int start = 0;
        int end = 0;
        unordered_map<char,int>mpp;
        while(end<s.length()){
            mpp[s[end]]++;
            while(mpp[s[end]]>1){
                mpp[s[start]]--;
                start++;
            }
            int currLen = end-start+1;
            maxi = max(maxi,currLen);
            end++;
        }
        return maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer>map = new HashMap<>();
        int maxi = 0;
        int start = 0;
        int end = 0;
        while(end<s.length()){
            char ch = s.charAt(end);
            map.put(ch,map.getOrDefault(ch,0)+1);
            while(map.get(ch)>1){
                char startChar = s.charAt(start);
                map.put(startChar,map.get(startChar)-1);
                start++;
            }
            maxi = Math.max(maxi,end-start+1);
            end++;
        }
        return maxi;
    }
}